Framework/Geometry/src/Objects/BeamlineRayTracer.cpp

```diff
@@ -0,0 +1,503 @@
+ * @param componentIndex :: The component to be checked.
+ */
+void checkIntersectionWithRectangularBank(Track &track,
+                                          const ComponentInfo &componentInfo,
+                                          size_t componentIndex) {
+
+  // Get the corners of the bank
+  auto corners = componentInfo.quadrilateralComponent(componentIndex);
+
+  // Store points of the bank
+  V3D bottomLeft = componentInfo.position(corners.bottomLeft);
+  V3D bottomRight = componentInfo.position(corners.bottomRight);
+  V3D topLeft = componentInfo.position(corners.topLeft);
+
+  // Horizontal (x-axis) basis vector of the detector
+  V3D horizontal = bottomRight - bottomLeft;
+
+  // Vertical (y-axis) basis vector of the detector
+  V3D vertical = topLeft - bottomLeft;
+
+  // The beam direction
+  V3D beam = track.direction();
+
+  // From: http://en.wikipedia.org/wiki/Line-plane_intersection (taken on May
+  // 4, 2011), We build a matrix to solve the linear equation:
+  Matrix<double> mat(3, 3);
+  mat.setColumn(0, beam * -1.0);
+  mat.setColumn(1, horizontal);
+  mat.setColumn(2, vertical);
+  mat.Invert();
+
+  // Multiply by the inverted matrix to find t,u,v
+  V3D tuv = mat * (track.startPoint() - bottomLeft);
+
+  // Intersection point
+  V3D intersec = beam;
+  intersec *= tuv[0];
+
+  // t = coordinate along the line
+  // u,v = coordinates along horizontal, vertical
+  // (and correct for it being between 0, xpixels-1).  The +0.5 is because the
+  // bottomLeft is at the CENTER of pixel 0,0.
+  double u = (double(corners.nX - 1) * tuv[1] + 0.5);
+  double v = (double(corners.nY - 1) * tuv[2] + 0.5);
+
+  // In indices
+  size_t xIndex = size_t(u);
+  size_t yIndex = size_t(v);
+
+  // Out of range?
+  if (xIndex >= corners.nX)
+    return;
+  if (yIndex >= corners.nY)
+    return;
+
+  // Get the componentIndex of the component in the assembly at the (X, Y)
+  // pixel position.
+  auto childrenX = componentInfo.children(componentIndex);
+  auto childrenY = componentInfo.children(childrenX[xIndex]);
+
+  // Create a link and add it to the track
+  track.addLink(intersec, intersec, 0.0, componentInfo.shape(childrenY[yIndex]),
+                componentInfo.componentID(childrenY[yIndex])->getComponentID());
+}
+
+void checkIntersectionWithOutlineComposite(Track &track,
+                                           const ComponentInfo &componentInfo,
+                                           size_t componentIndex) {
+
+  // Get the corners of the bank
+  ComponentInfo::QuadrilateralComponent corners;
+  const auto &innerRangeComp = componentInfo.children(componentIndex);
+  // nSubComponents, subtract off self hence -1. nSubComponents = number of
+  // horizontal columns.
+  corners.nY = innerRangeComp.size();
+  corners.nX = 0;
+
+  corners.bottomLeft = innerRangeComp.front();
+  corners.topRight = innerRangeComp.back();
+  corners.topLeft = innerRangeComp.back();
+  corners.bottomRight = innerRangeComp.front();
```

Comment:
It is a little weird to see it extracted like this. i.e treating a tube (effectively as line as far as centre positions are defined) like a quad. That's just an observation, the rest of it looks ok.